### PR TITLE
Add HunyuanVideo T2V Preprocessing and Training Support

### PR DESCRIPTION
**Pull request by henryfzh**
### This is a PR adds training and preprocessing support for HunyuanVideo T2V models

The example usage of the preprocessing and training pipeline is in `fastvideo/tests/nightly/test_e2e_hunyuan_overfit_single_sample.py`

**New Files:**
`fastvideo/training/hunyuan_training_pipeline.py` - HunyuanVideo training pipeline
`fastvideo/pipelines/preprocess/hunyuan/hunyuan_preprocess_pipelins.py` - Preprocessing pipelines for T2V
`fastvideo/tests/nightly/test_e2e_hunyuan_overfit_single_sample.py` - E2E overfit test
Example scripts in `examples/training/finetune/hunyuan_t2v/ and hunyuan_i2v/`

**Changes:**
`fastvideo/configs/models/encoders/clip.py & llama.py `- Added tokenizer kwargs
`fastvideo/dataset/dataloader/record_schema.py` - Handle models without image encoders
`fastvideo/dataset/utils.py` - Added dtype support for tensor loading
`fastvideo/layers/rotary_embedding.py` - Fixed tensor reshaping
`fastvideo/models/vaes/hunyuanvae.py` - Fixed attention mask expansion
`fastvideo/training/training_pipeline.py` - Auto-detect model type for normalization
`fastvideo/workflow/preprocess/components.py` - Auto-calculate num_frames from fps/duration

### CODE REVIEW

File: examples/training/finetune/hunyuan_i2v/download_dataset.sh [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+python scripts/huggingface/download_hf.py --repo_id "wlsaidhi/crush-smol-merged" --local_dir "data/crush-smol" --repo_type "dataset"
```
Comment (SolitaryThinker) [inline examples/training/finetune/hunyuan_i2v/download_dataset.sh]: add newline char

File: examples/training/finetune/hunyuan_t2v/download_dataset.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+python scripts/huggingface/download_hf.py --repo_id "wlsaidhi/crush-smol-merged" --local_dir "data/crush-smol" --repo_type "dataset"
```
Comment (SolitaryThinker) [inline examples/training/finetune/hunyuan_t2v/download_dataset.sh]: same here

File: fastvideo/configs/models/encoders/clip.py [hunk 3]
```diff
@@ -80,6 +80,14 @@ class CLIPTextConfig(TextEncoderConfig):
...
     is_causal: bool = True
     prefix: str = "clip"
+    tokenizer_kwargs: dict = field(
+        default_factory=lambda: {
+            "padding": "max_length",
+            "truncation": True,
+            "max_length": 77,
+            "return_tensors": "pt",
+        }
+    )
```
Comment (SolitaryThinker) [inline fastvideo/configs/models/encoders/clip.py]: this should be put in `CLIPTextArchConfig` instead. For an example you can look at the T5 encoder

File: fastvideo/configs/models/encoders/llama.py [hunk 4]
```diff
@@ -62,3 +62,11 @@ class LlamaConfig(TextEncoderConfig):
...
 
     prefix: str = "llama"
+    tokenizer_kwargs: dict = field(
+        default_factory=lambda: {
+            "padding": "max_length",
+            "truncation": True,
+            "max_length": 256,
+            "return_tensors": "pt",
+        }
+    )
```
Comment (SolitaryThinker) [inline fastvideo/configs/models/encoders/llama.py:66]: same here, should go in the archconfig

File: fastvideo/dataset/dataloader/record_schema.py [hunk 5]
```diff
@@ -72,9 +72,11 @@ def i2v_record_creator(batch: PreprocessBatch) -> list[dict[str, Any]]:
...
     records = basic_t2v_record_creator(batch)
 
-    assert len(
-        batch.image_embeds) == 1, "image embedding should be a single tensor"
-    image_embeds = batch.image_embeds[0]
+    # Adapt for model doesn't have image encoder, e.g., Hunyuan
+    if len(batch.image_embeds) == 0:
+        image_embeds = None
+    elif len(batch.image_embeds) == 1:
+        image_embeds = batch.image_embeds[0]
```
Comment (SolitaryThinker) [inline fastvideo/dataset/dataloader/record_schema.py:79]: could we add another else and raise an exception here?

File: fastvideo/layers/rotary_embedding.py [hunk 6]
```diff
@@ -194,14 +194,14 @@ def forward_native(
...
-        query = query.view(num_tokens, -1, self.head_size)
+        query = query.reshape(num_tokens, -1, self.head_size)
         query_rot = query[..., :self.rotary_dim]
         query_pass = query[..., self.rotary_dim:]
         query_rot = _apply_rotary_emb(query_rot, cos, sin, self.is_neox_style)
         query = torch.cat((query_rot, query_pass), dim=-1).reshape(query_shape)
 
         key_shape = key.shape
-        key = key.view(num_tokens, -1, self.head_size)
+        key = key.reshape(num_tokens, -1, self.head_size)
```
Comment (SolitaryThinker) [inline fastvideo/layers/rotary_embedding.py:204]: why is the change from view to reshape needed?

File: fastvideo/training/hunyuan_training_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+class HunyuanTrainingPipeline(TrainingPipeline):
+    """
+    A training pipeline for Hunyuan.
+    """
+    _required_config_modules = ["scheduler", "transformer", "vae"]
+
+    def initialize_pipeline(self, fastvideo_args: FastVideoArgs):
+        self.modules["scheduler"] = FlowMatchEulerDiscreteScheduler(
+            shift=fastvideo_args.pipeline_config.flow_shift)
```
Comment (SolitaryThinker) [inline fastvideo/training/hunyuan_training_pipeline.py:26]: double check this, this is a special case for wan and hunyuan video may use a different scheduler. Let me know if you don't know where to find this info.

File: fastvideo/layers/rotary_embedding.py (hunk 6 above)
Comment (henryfzh) [reply fastvideo/layers/rotary_embedding.py:204]: Because when key or query are empty, view() will raise a runtime error.

File: fastvideo/dataset/dataloader/record_schema.py (hunk 5 above)
Comment (henryfzh) [reply fastvideo/dataset/dataloader/record_schema.py:79]: ok, will do.

**Comment by SolitaryThinker** [comment]
Hi could you run pre-commit, thanks.

**Comment by SolitaryThinker** [comment]
Closing as superseded by the modular Hunyuan training support merged in #1175, with LoRA support in #1242. This branch targets the legacy training stack and now conflicts with current main; making it mergeable would require a redesign rather than a rebase. If any generic preprocessing behavior here is still wanted, please send it as a focused current-main PR with runnable preprocessing and component coverage. Thanks for preserving the original requirements and discussion here.
